File: esp32/scanner/main/detection/french_dri_parser.c
/**
 * Friend or Foe -- French "Signalement Electronique a Distance" Parser (ESP32-S3)
 *
 * Ported from Android FrenchDriParser.kt.
 * Parses French DRI vendor-specific IEs from WiFi beacons using
 * OUI 6A:5C:35, VS type 0x01, with TLV-encoded UTF-8 fields.
 *
 * TLV types:
 *   1  = Protocol version
 *   2  = ID_FR (30-char French identifier: trigram + model + serial)
 *   3  = ANSI/CTA-2063 ID (physical serial number)
 *   4  = Latitude (ASCII decimal degrees)
 *   5  = Longitude (ASCII decimal degrees)
 *   6  = Altitude MSL (ASCII meters)
 *   7  = Height AGL (ASCII meters)
 *   8  = Takeoff latitude (ASCII decimal degrees)
 *   9  = Takeoff longitude (ASCII decimal degrees)
 *   10 = Ground speed (ASCII m/s)
 *   11 = Heading (ASCII degrees 0-359)
 */

#include "french_dri_parser.h"
#include "constants.h"

#include <string.h>
#include <stdlib.h>
#include <esp_log.h>

static const char *TAG = "french_dri";

/* French DRI OUI: 6A:5C:35 */
#define FRENCH_OUI_0    0x6A
#define FRENCH_OUI_1    0x5C
#define FRENCH_OUI_2    0x35
#define FRENCH_VS_TYPE  0x01

/* Minimum payload: 3 OUI + 1 type + 3 (one minimal TLV) = 7 bytes */
#define FRENCH_MIN_PAYLOAD_SIZE  7

/* TLV type codes */
#define TLV_PROTOCOL_VERSION  1
#define TLV_ID_FR             2
#define TLV_ANSI_CTA_ID       3
#define TLV_LATITUDE          4
#define TLV_LONGITUDE         5
#define TLV_ALTITUDE_MSL      6
#define TLV_HEIGHT_AGL        7
#define TLV_TAKEOFF_LAT       8
#define TLV_TAKEOFF_LON       9
#define TLV_GROUND_SPEED      10
#define TLV_HEADING           11

/* ── Helper: safe string-to-double from non-null-terminated buffer ────────── */

static double parse_ascii_double(const uint8_t *buf, size_t len)
{
    /* Copy to a null-terminated stack buffer (max 32 chars) */
    char tmp[33];
    size_t n = (len < sizeof(tmp) - 1) ? len : sizeof(tmp) - 1;
    memcpy(tmp, buf, n);
    tmp[n] = '\0';
    return atof(tmp);
}

static float parse_ascii_float(const uint8_t *buf, size_t len)
{
    return (float)parse_ascii_double(buf, len);
}

/* ── Public API ─────────────────────────────────────────────────────────── */

bool french_dri_parse_ie(const uint8_t *payload, size_t len, odid_state_t *state)
{
    if (payload == NULL || state == NULL) return false;
    if (len < FRENCH_MIN_PAYLOAD_SIZE) return false;

    /* Verify French OUI */
    if (payload[0] != FRENCH_OUI_0 ||
        payload[1] != FRENCH_OUI_1 ||
        payload[2] != FRENCH_OUI_2) {
        return false;
    }

    /* Verify VS type */
    if (payload[3] != FRENCH_VS_TYPE) {
        return false;
    }

    /* Parse TLV fields starting at offset 4 */
    size_t offset = 4;
    bool parsed = false;

    while (offset + 2 <= len) {
        uint8_t tlv_type = payload[offset];
        uint8_t tlv_len  = payload[offset + 1];
        offset += 2;

        if (offset + tlv_len > len) break;

        const uint8_t *value = &payload[offset];

        switch (tlv_type) {
        case TLV_ID_FR:
            if (tlv_len > 0 && tlv_len < sizeof(state->drone_id)) {
                memcpy(state->drone_id, value, tlv_len);
                state->drone_id[tlv_len] = '\0';
                parsed = true;
            }
            break;

        case TLV_ANSI_CTA_ID:
            /* Use as drone_id if ID_FR not yet set */
            if (state->drone_id[0] == '\0' && tlv_len > 0 &&
                tlv_len < sizeof(state->drone_id)) {
                memcpy(state->drone_id, value, tlv_len);
                state->drone_id[tlv_len] = '\0';
            }
            break;

        case TLV_LATITUDE: {
            double lat = parse_ascii_double(value, tlv_len);
            if (lat >= -90.0 && lat <= 90.0 && lat != 0.0) {
                state->latitude = lat;
                state->has_location = true;
                parsed = true;
            }
            break;
        }

        case TLV_LONGITUDE: {
            double lon = parse_ascii_double(value, tlv_len);
            if (lon >= -180.0 && lon <= 180.0 && lon != 0.0) {
                state->longitude = lon;
                state->has_location = true;
                parsed = true;
            }
            break;
        }

        case TLV_ALTITUDE_MSL:
            state->altitude_m = parse_ascii_double(value, tlv_len);
            break;

        case TLV_HEIGHT_AGL:
            state->height_agl_m = parse_ascii_double(value, tlv_len);
            break;

        case TLV_TAKEOFF_LAT: {
            double lat = parse_ascii_double(value, tlv_len);
            if (lat >= -90.0 && lat <= 90.0 && lat != 0.0) {
                state->operator_lat = lat;
                /* operator location set */
            }
            break;
        }

        case TLV_TAKEOFF_LON: {
            double lon = parse_ascii_double(value, tlv_len);
            if (lon >= -180.0 && lon <= 180.0 && lon != 0.0) {
                state->operator_lon = lon;
                /* operator location set */
            }
            break;
        }

        case TLV_GROUND_SPEED:
            state->speed_mps = parse_ascii_float(value, tlv_len);
            break;

        case TLV_HEADING: {
            float hdg = parse_ascii_float(value, tlv_len);
            if (hdg >= 0.0f && hdg <= 360.0f) {
                state->heading_deg = hdg;
            }
            break;
        }

        default:
            /* Unknown TLV type — skip */
            break;
        }

        offset += tlv_len;
    }

    if (parsed) {
        ESP_LOGD(TAG, "Parsed French DRI: id=%s lat=%.6f lon=%.6f alt=%.0fm",
                 state->drone_id, state->latitude, state->longitude,
                 state->altitude_m);
    }

    return parsed;
}

bool french_dri_parse_frame(const uint8_t *frame, size_t len, odid_state_t *state)
{
    if (frame == NULL || state == NULL) return false;

    /* Tagged parameters start after MAC header (24) + fixed fields (12) */
    const size_t TAGGED_PARAMS_OFFSET = 24 + 12;

    if (len < TAGGED_PARAMS_OFFSET + 2) return false;

    size_t pos = TAGGED_PARAMS_OFFSET;
    bool found = false;

    while (pos + 2 <= len) {
        uint8_t tag_id  = frame[pos];
        uint8_t tag_len = frame[pos + 1];
        size_t ie_start = pos + 2;

        if (ie_start + tag_len > len) break;

        if (tag_id == IE_VENDOR_SPECIFIC && tag_len >= FRENCH_MIN_PAYLOAD_SIZE) {
            if (french_dri_parse_ie(&frame[ie_start], tag_len, state)) {
                found = true;
            }
        }

        pos = ie_start + tag_len;
    }

    return found;
}

File: android/app/src/main/java/com/friendorfoe/detection/FrenchDriParser.kt
package com.friendorfoe.detection

import android.net.wifi.ScanResult
import android.os.Build
import android.util.Log
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Parser for French "Signalement Electronique a Distance" (Electronic Remote Identification)
 * per Arrete du 27 decembre 2019.
 *
 * French drones broadcast identification and telemetry in WiFi beacon vendor-specific
 * Information Elements using OUI 6A:5C:35, VS type 0x01. The payload uses TLV
 * (Type-Length-Value) encoding with the following fields:
 *
 * | Type | Field              | Encoding                          |
 * |------|--------------------|-----------------------------------|
 * | 1    | Protocol version   | uint8                             |
 * | 2    | ID_FR              | 30-char UTF-8 (trigram+model+SN)  |
 * | 3    | ANSI/CTA-2063 ID   | UTF-8 physical serial number      |
 * | 4    | Latitude           | ASCII decimal degrees [-90, +90]  |
 * | 5    | Longitude          | ASCII decimal degrees (-180, +180]|
 * | 6    | Altitude MSL       | ASCII meters                      |
 * | 7    | Height AGL         | ASCII meters                      |
 * | 8    | Takeoff latitude   | ASCII decimal degrees             |
 * | 9    | Takeoff longitude  | ASCII decimal degrees             |
 * | 10   | Ground speed       | ASCII m/s                         |
 * | 11   | Heading            | ASCII degrees 0-359               |
 *
 * Requires API 33+ for [ScanResult.getInformationElements].
 */
object FrenchDriParser {

    private const val TAG = "FrenchDriParser"

    /** French DRI OUI: 6A:5C:35 */
    private val FRENCH_OUI = byteArrayOf(0x6A, 0x5C, 0x35)

    /** VS type byte for French DRI protocol */
    private const val VS_TYPE: Byte = 0x01

    /** Vendor-specific IE element ID per IEEE 802.11 */
    private const val IE_VENDOR_SPECIFIC = 221

    /** Minimum IE payload: 3 (OUI) + 1 (type) + at least 1 TLV = 7 */
    private const val MIN_PAYLOAD_SIZE = 7

    // TLV type codes
    private const val TLV_PROTOCOL_VERSION = 1
    private const val TLV_ID_FR = 2
    private const val TLV_ANSI_CTA_ID = 3
    private const val TLV_LATITUDE = 4
    private const val TLV_LONGITUDE = 5
    private const val TLV_ALTITUDE_MSL = 6
    private const val TLV_HEIGHT_AGL = 7
    private const val TLV_TAKEOFF_LAT = 8
    private const val TLV_TAKEOFF_LON = 9
    private const val TLV_GROUND_SPEED = 10
    private const val TLV_HEADING = 11

    /**
     * Attempt to parse French DRI data from a WiFi scan result.
     *
     * @param scanResult WiFi scan result to inspect for French DRI vendor IEs
     * @param state Mutable accumulator for partial drone state (reused from ODID)
     * @return true if French DRI data was successfully parsed
     */
    fun parse(scanResult: ScanResult, state: OpenDroneIdParser.DronePartialState): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return false
        }

        return try {
            parseApi33(scanResult, state)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to parse French DRI IE: ${e.message}")
            false
        }
    }

    @Suppress("NewApi")
    private fun parseApi33(
        scanResult: ScanResult,
        state: OpenDroneIdParser.DronePartialState
    ): Boolean {
        val ies = scanResult.informationElements ?: return false

        for (ie in ies) {
            if (ie.id != IE_VENDOR_SPECIFIC) continue

            val bytes = ie.bytes ?: continue
            val payload = ByteArray(bytes.remaining())
            bytes.get(payload)
            bytes.rewind()

            if (payload.size < MIN_PAYLOAD_SIZE) continue

            // Check French OUI
            if (payload[0] != FRENCH_OUI[0] ||
                payload[1] != FRENCH_OUI[1] ||
                payload[2] != FRENCH_OUI[2]
            ) continue

            // Check VS type
            if (payload[3] != VS_TYPE) continue

            // Parse TLV fields starting at offset 4
            return parseTlvPayload(payload, 4, state)
        }

        return false
    }

    /**
     * Parse TLV-encoded fields from the French DRI payload.
     */
    private fun parseTlvPayload(
        payload: ByteArray,
        startOffset: Int,
        state: OpenDroneIdParser.DronePartialState
    ): Boolean {
        var offset = startOffset
        var parsed = false

        while (offset + 2 <= payload.size) {
            val tlvType = payload[offset].toInt() and 0xFF
            val tlvLength = payload[offset + 1].toInt() and 0xFF
            offset += 2

            if (offset + tlvLength > payload.size) break

            val valueBytes = payload.copyOfRange(offset, offset + tlvLength)
            val valueStr = String(valueBytes, Charsets.UTF_8).trim()

            when (tlvType) {
                TLV_PROTOCOL_VERSION -> {
                    // Protocol version, informational only
                }
                TLV_ID_FR -> {
                    if (valueStr.isNotEmpty()) {
                        state.droneId = valueStr
                        parsed = true
                    }
                }
                TLV_ANSI_CTA_ID -> {
                    // Physical serial number — use as droneId if ID_FR is absent
                    if (state.droneId == null && valueStr.isNotEmpty()) {
                        state.droneId = valueStr
                    }
                }
                TLV_LATITUDE -> {
                    valueStr.toDoubleOrNull()?.let { lat ->
                        if (lat in -90.0..90.0 && lat != 0.0) {
                            state.latitude = lat
                            parsed = true
                        }
                    }
                }
                TLV_LONGITUDE -> {
                    valueStr.toDoubleOrNull()?.let { lon ->
                        if (lon in -180.0..180.0 && lon != 0.0) {
                            state.longitude = lon
                            parsed = true
                        }
                    }
                }
                TLV_ALTITUDE_MSL -> {
                    valueStr.toDoubleOrNull()?.let { alt ->
                        state.altitudeMeters = alt
                    }
                }
                TLV_HEIGHT_AGL -> {
                    valueStr.toDoubleOrNull()?.let { height ->
                        state.heightAglMeters = height
                    }
                }
                TLV_TAKEOFF_LAT -> {
                    valueStr.toDoubleOrNull()?.let { lat ->
                        if (lat in -90.0..90.0 && lat != 0.0) {
                            state.operatorLatitude = lat
                        }
                    }
                }
                TLV_TAKEOFF_LON -> {
                    valueStr.toDoubleOrNull()?.let { lon ->
                        if (lon in -180.0..180.0 && lon != 0.0) {
                            state.operatorLongitude = lon
                        }
                    }
                }
                TLV_GROUND_SPEED -> {
                    valueStr.toFloatOrNull()?.let { speed ->
                        state.speedMps = speed
                    }
                }
                TLV_HEADING -> {
                    valueStr.toFloatOrNull()?.let { heading ->
                        if (heading in 0f..360f) {
                            state.heading = heading
                        }
                    }
                }
            }

            offset += tlvLength
        }

        if (parsed) {
            Log.d(TAG, "Parsed French DRI: id=${state.droneId}, " +
                    "lat=${state.latitude}, lon=${state.longitude}, " +
                    "alt=${state.altitudeMeters}m")
        }

        return parsed
    }
}
